# ASVCLR
Accurate Structural Variation Caller for Long Reads

-------------------
ASVCLR is an accurate Structural Variation Caller for Long Reads, such as PacBio sequencing and Oxford Nanopore sequencing. ASVCLR can detect both short indels (e.g. <50 bp) and long structural varitions (e.g. >=50 bp), such as duplications, inversions and translocations, and producing fewer false positives with high precise variant margins.  

## Prerequisites
ASVCLR depends on the following libraries and tools:
* HTSlib (http://www.htslib.org/download/)
* Canu (v1.7.1) (https://github.com/marbl/canu/releases)
* BLAT (http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/blat/)
* g++.

The above library and tools should be installed before compiling ASVCLR. Canu and BLAT should be globally accessible in the computer system, these executable files `canu` and `blat` should be placed or linked to the `$PATH` directory.

Note that: Canu v1.7.1 is about 5 to 10 folds faster than Canu v1.8, but it may cannot construct the assembly results (contigs) in some genomic regions as the overlap failure, therefore if you care more about the accuracy of the results than the running time, please use Canu v1.8 instead.

According to our human chromosome 1 simulated results, the Recall was increased from xxx to xxx, the Precision was increased from xxx to xxx, and the F1 score was increased from xxx to xxx when using Canu v1.8 instead of v.1.7.1, which is a very small improvement.


## Compiling ASVCLR

The binary file can be generated by typing:
```sh
$ ./autogen.sh
```
and the binary file 'asvclr' will be output into the folder 'bin' in the package directory.


## Quick Usage

Simply, ASVCLR can be run by typing the `all` command:
```sh
$ asvclr all -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
```
Then, the following commands `detect`, `assemble` and `call` will be performed in turn. The help information can be shown:
```sh
$ asvclr
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 0.1.0 (using htslib 1.9)

Usage:  asvclr  <command> [options]

Commands:
     detect       detect indel signatures in aligned reads
     assemble     assemble candidate regions and align assemblies
                  back to reference
     call         call indels by alignments of local genome assemblies
     all          run the above commands in turn
```


## Usage

Alternatively, there are three steps to run ASVCLR: `detect`, `assemble` and `call`.

```sh
$ asvclr detect -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
$ asvclr assemble -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
$ asvclr call -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
```

The reference and an sorted BAM file will be the input of ASVCLR, and the variants stored in the VCF file format and translocations in BEDPE file format will be generated as the output.


### `Detect` Step
 
Structural variant regions will be detected according to abnormal features. These regions includes insertions, deletions, duplications, inversions and translocations.

```sh
$ asvclr detect -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr detect
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 0.1.0 (using htslib 1.9)

Usage: asvclr detect [options] <in.bam>|<in.sam>

Options: 
     -f FILE      reference file name (required)
     -b INT       block size [1000000]
     -s INT       detect slide size [500]
     -m INT       minimal SV size to detect [2]
     -n INT       minimal clipping reads supporting a SV [7]
     -c INT       maximal clipping region size to detect [10000]
     -o FILE      prefix of the output file [stdout]
     -t INT       number of threads [0]
     -M INT       Mask mis-aligned regions [1]: 1 for yes, 0 for no
     -h           show this help message and exit
```

### `Assemble` Step

Perform local assembly for the detected variant regions using Canu, and extract the corresponding local reference.

```sh
$ asvclr assemble -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr assemble
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 0.1.0 (using htslib 1.9)

Usage: asvclr assemble [options] <in.bam>|<in.sam>

Options: 
     -f FILE      reference file name (required)
     -b INT       block size [1000000]
     -S INT       assemble slide size [10000]
     -c INT       maximal clipping region size to assemble [10000]
     -o FILE      prefix of the output file [stdout]
     -t INT       number of threads [0]
     -M INT       Mask mis-aligned regions [1]: 1 for yes, 0 for no
     -h           show this help message and exit

```

### `Call` Step

Align the assembly result (contigs) to its local reference using BLAT to generate the sim4 formated alignments, and call each type variations using the BLAT alignments.

```sh
$ asvclr call -t 14 -c 20000 -f hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr call
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 0.1.0 (using htslib 1.9)

Usage: asvclr call [options] <in.bam>|<in.sam>

Options: 
     -f FILE      reference file name (required)
     -b INT       block size [1000000]
     -S INT       assemble slide size [10000]
     -c INT       maximal clipping region size to call [10000]
     -o FILE      prefix of the output file [stdout]
     -t INT       number of threads [0]
     -M INT       Mask mis-aligned regions [1]: 1 for yes, 0 for no
     -h           show this help message and exit
```


## Output Result Description

There are two kinds of output files: VCF file and BEDPE file. Insertions, deletions, inversions and duplications are stored in the VCF file format, and translocations are saved in the BEDPE file format.



------------------
If you have problems or some suggestions, please contact: xzhu@hrbnu.edu.cn without hesitation. 

---- Enjoy !!! -----

