# ASVCLR

```
     @@@            @@@@@         @@     @@@          @@@@@         @@@            @@@@@@  
     @@@           @@@@@@@        @@@    @@@        @@@@@@@@        @@@            @@@@@@@ 
    @@@@@          @@@            @@@   @@@         @@@             @@@            @@   @@ 
    @@@@@          @@@            @@@   @@@        @@@              @@@            @@   @@ 
   @@@ @@          @@@@            @@@ @@@         @@@              @@@            @@  @@@ 
   @@@ @@@          @@@@@          @@@ @@@         @@@              @@@            @@@@@@  
   @@@@@@@           @@@@@         @@@ @@@         @@@              @@@            @@@@@@  
  @@@@@@@@@            @@@          @@@@@          @@@              @@@            @@  @@  
  @@@   @@@        @   @@@          @@@@@           @@@   @         @@@            @@  @@@ 
 @@@    @@@        @@@@@@@           @@@            @@@@@@@@        @@@@@@@        @@   @@@
 @@@     @@@        @@@@@            @@@              @@@@@         @@@@@@@        @@   @@@
```

Allele-aware Structural Variant Caller for Long Reads

-------------------
ASVCLR is an Allele-aware Structural Variant (SV) Caller for Long Reads, such as PacBio sequencing and Oxford Nanopore sequencing. ASVCLR can detect both simple indels (insertions and deletions) and the hard-to-call complex structural varitions (SVs) (e.g. >=20 bp), such as tandem duplications, inversions, and producing fewer false positives with high recall and precision for various long-read sequencing technologies.

The main features of ASVCLR can be summarized as follows:
* ASVCLR detects candidate variant regions according to the variant signatures and categorizes these regions as indel regions and clipping regions. The indel regions typically have much more indel events within the intra-aligned segments whereas the clipping regions typically have much more clipping events (soft- and hard-clippings). ASVCLR was designed to infer the variants in the indel regions and clipping regions, respectively.
* ASVCLR consists of three steps (or commands): `det`, `cns` and `call`. In `det` (detect) step, the indel regions and clipping regions are detected as candidate variant regions; and subsequently, in the `cns` (consensus) step, ASVCLR performs the consensus operation to generate the consensus sequences by using abPOA and wtdbg2 for these candidate indel regions and clipping regions, respectively; and finally, in the `call` step, these consensus sequences derived from indel regions and clipping regions are aligned back to local reference by using minimap2 to infer the variant locations and sequences.
* ASVCLR extracts variant features in indel regions and clipping regions, respectively, and adopts an allele-aware clustering for these regions to avoid the interfere impact of different alleles to generate more accurate variant call results.

For more detailed experiment information, please refer to [asvclr-experiments](https://github.com/zhuxiao/asvclr-experiments/).

## Prerequisites
ASVCLR depends on the following libraries and tools:
* HTSlib (http://www.htslib.org/download/)
* abPOA (https://github.com/yangao07/abPOA/releases)
* minimap2 (https://github.com/lh3/minimap2/releases)
* g++ (v4.7 or higher which supports c++11)
* wtdbg2 2.5 (https://github.com/ruanjue/wtdbg2/releases)
The above library and tools should be installed before compiling ASVCLR. abPOA, minimap2 and wtdbg2 should be globally accessible in the computer system, these executable files `abpoa`, `minimap2`, `wtdbg2.pl` and `wtdbg2` or their soft links should be included in the `$PATH` directory.


## Compiling ASVCLR

The binary file can be generated by typing:
```sh
$ wget -c https://github.com/zhuxiao/asvclr/releases/download/1.4.6/asvclr_1.4.6.tar.xz
$ tar -xf asvclr_1.4.6.tar.xz
$ cd asvclr_1.4.6/
$ ./auto_gen.sh
```
And the binary file `asvclr` will be output into the folder `bin` in this package directory.

Besides, it is recommanded to create the link of `asvclr` in `bin` directory to the `$PATH` directory.
```sh
# create soft-link of asvclr to '~/bin' which should be already exist or create a new one
$ mkdir ~/bin
$ ln -s /path/to/asvclr/bin/asvclr ~/bin
# or, create soft-link of asvclr to /usr/local/bin, which needs the sudo permission
$ sudo ln -s /path/to/asvclr/bin/asvclr /usr/local/bin
```


### Compiling Prerequisites

Compling prerequisites:

* Ubuntu:
```sh
$ sudo apt install g++
```

* CentOS:
```sh
$ sudo dnf install gcc-c++
```


## Quick Start

Simply, ASVCLR can be run by typing the `all` command:
```sh
$ asvclr all -t 32 -n 3 -m 20 -x ccs -p asvclr -o out_dir ref.fa genome_sorted.bam
```
Then, the following commands `det`, `cns`, `call`, `all` and `det-cns` will be performed in turn.

Moreover, ASVCLR can be used to detect variants for some user-specified regions.
```sh
# call variants in user-specified regions of reference: chr1, chr2:10000000-20000000, chr3:50000000-100000000
$ asvclr all -t 32 -n 3 -m 20 -x ccs -p asvclr -o out_dir ref.fa genome_sorted.bam chr1 chr2:10000000-20000000 chr3:50000000-100000000
```
Then, only the three regions `chr1`, `chr2:10000000-20000000` and `chr3:50000000-100000000` will be processed.

The help information can be shown:
```sh
$ asvclr all
Program: ASVCLR (Allele-aware Structural Variant Caller for Long Reads)
Version: 1.4.6 (using htslib 1.17)

Usage: asvclr all [options] <REF_FILE> <BAM_FILE> [Region ...]

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted file (required)
   Region        Reference regions to process: CHR|CHR:START-END.
                 If unspecified, all reference regions will be 
                 processed (optional)

Options: 
   -b INT        block size [1000000]
   -s INT        Slide window size [500]
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be max(2+floor(0.03*depth), 3).
   -x STR        sequencing technology preset: rs, ont, sq, ccs. [ccs]
                         ccs: -i 0.9 -I 0.95 -N 0.1
                   sq/rs/ont: -i 0.75 -I 0.7 -N 0.2
   -i FLOAT      minimal sequence identity for variant match. [0.9]
   -I FLOAT      minimal sequence identity for merge [0.95]
   -d INT        minimal reference distance for merge [300]
                 Neighboring SVs with distance smaller than INT and sequence identity
                 larger than FLOAT will be merged
   -r FLOAT      minimal ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.5]
   -N FLOAT      maximal NM ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.1]
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -q INT        minimal mapping quality [20]
                 Reads with mapping quality smaller than threshold will be ignored
   -c FLOAT      expected sampling coverage for local consensus [40], 
                 0 for no coverage sampling
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --cns-chunk-size INT
                 maximal reference chunk size to collect reads data to perform
                 local consensus [1000]. Reads of variants with reference
                 distance < INT will be collected to perform local consensus
   --cns-side-ext-size INT
                 region extend size on both sides while generating consensus sequences
                 [1000] bp.
   --cns-side-ext-size-clip INT
                 clip region extend size on both sides while generating consensus sequences
                 [5000] bp.
   --min-cons-read-size INT
                 minimal segment size for consensus [100] bp.
   --keep-cns-reads
                 Keep temporary reads from being deleted during local consensus.
                 This may take some additional disk space
   --re-cns-failed-work
                 Reperform previously failed local consensus work.
   --include-alt
                 include alt chromosomal items in result [False]
   --include-decoy
                 include decoy chromosomal items in result [False]
   --sample STR  Sample name ["sample"]
   --gt-min-consist-merge FLOAT
                 minimal sequence identity for allele merge [0.95].
                 Allelic Variants will be merged if their sequence identity are larger than FLOAT. 
   --gt-homo-ratio FLOAT
                 minimal allele ratio for homozygous alleles [0.75].
                 Variant is homozygous if the ratio of allele count is larger than FLOAT.
   --gt_hete_ratio FLOAT
                 minimal allele ratio for heterozygous alleles [0.2].
                 Variant is heterozygous if the ratio of allele count is larger than FLOAT.
   -v,--version  show version information
   -h,--help     show this help message and exit

Example:
   # run the pipeline on the whole genome for PacBio CCS sequencing
   $ asvclr all -t 32 -n 3 -m 20 -x ccs -o output ref.fa genome_sorted.bam

   # run the pipeline to analyze the user-specified regions: chr1, chr2:10000000-20000000
   $ asvclr all -t 32 -n 3 -m 20 -x ccs -o output ref.fa genome_sorted.bam chr1 chr2:10000000-20000000
```
where, the htslib version is the version of HTSlib installed on the machine.

Besides, the overall help information can be shown as below:

```sh
$ asvclr
Program: ASVCLR (Allele-aware Structural Variant Caller for Long Reads)
Version: 1.4.6 (using htslib 1.17)

Usage:  asvclr <command> [options] <REF_FILE> <BAM_FILE> [Region ...]

Description:
   REF_FILE          Reference file (required)
   BAM_FILE          Coordinate-sorted BAM file (required)
   Region            Reference regions to process: CHR|CHR:START-END.
                     If unspecified, all reference regions will be 
                     processed (optional)

Commands:
   det               detect indel signatures in aligned reads
   cns               consensus candidate regions
   call              call indels by alignments of local consensus
   all               run the above commands in turn
   det-cns           run 'det' and 'cns' commands in turn

Example:
   # run the pipeline on the whole genome for PacBio CCS sequencing
   $ asvclr all -t 32 -n 3 -m 20 -x ccs -o output ref.fa genome_sorted.bam

   # run the pipeline to analyze user-specified regions: chr1, chr2:10000000-20000000
   $ asvclr all -t 32 -n 3 -m 20 -x ccs -o output ref.fa genome_sorted.bam chr1 chr2:10000000-20000000
```


## Usage

Alternatively, there are three steps to run ASVCLR: `det`, `cns` and `call`.

```sh
$ asvclr det -t 32 -n 3 -m 20 -o out_dir ref.fa genome_sorted.bam
$ asvclr cns -t 32 -x ccs -n 3 -m 20 -o out_dir ref.fa genome_sorted.bam
$ asvclr call -t 32 -x ccs -n 3 -m 20 -p asvclr -o out_dir ref.fa genome_sorted.bam
```

The reference and an sorted BAM file will be the input of ASVCLR, and the variants stored in the VCF file format will be generated as the output.


### `Det` Step

Structural variant regions will be detected according to variant signatures. These regions includes insertions, deletions, duplications, inversions and translocations.

```sh
$ asvclr det -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr det
Program: ASVCLR (Allele-aware Structural Variant Caller for Long Reads)
Version: 1.4.6 (using htslib 1.17)

Usage: asvclr det [options] <REF_FILE> <BAM_FILE> [Region ...]

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted BAM file (required)
   Region        Reference regions to process: CHR|CHR:START-END.
                 If unspecified, all reference regions will be 
                 processed (optional)

Options: 
   -b INT        block size [1000000]
   -s INT        Slide window size [500]
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be max(2+floor(0.03*depth), 3).
   -N FLOAT      maximal NM ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.1]
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -q INT        minimal mapping quality [20]
                 Reads with mapping quality smaller than threshold will be ignored
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --min-cons-read-size INT
                 minimal segment size for consensus [100] bp.
   --include-alt
                 include alt chromosomal items in result [False]
   --include-decoy
                 include decoy chromosomal items in result [False]
   --sample STR  Sample name ["sample"]
   -v,--version  show version information
   -h,--help     show this help message and exit

Example:
   # run 'det' command on the whole genome
   $ asvclr det -t 32 -n 3 -m 20 -o output ref.fa genome_sorted.bam

   # run 'det' command to analyze the user-specified regions: chr1, chr2:10000000-20000000
   $ asvclr det -t 32 -n 3 -m 20 -o output ref.fa genome_sorted.bam chr1 chr2:10000000-20000000
```

### `Cns` Step

In `Cns` step, ASVCLR extracts the corresponding local reference first. And then, ASVCLR generate the contigs by different strategies.
* For Indels in the inner-alignments, ASVCLR clusters the variant signatures by the modified NW (Needleman Wunsch) pairwise alignment algorithm, and for each cluster, ASVCLR extracts the read sequences and smoothing to remove sequence errors and short variant items. Then, ASVCLR performs the consensus operation by using abPOA to generate consensus sequences.
* For variants in clipping regions, such as DUPs, INVs, ASVCLR performs local assembly by using wtdbg2.

```sh
$ asvclr cns -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr cns
Program: ASVCLR (Allele-aware Structural Variant Caller for Long Reads)
Version: 1.4.6 (using htslib 1.17)

Usage: asvclr cns [options] <REF_FILE> <BAM_FILE>

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted BAM file (required)

Options: 
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be max(2+floor(0.03*depth), 3).
   -x STR        sequencing technology preset: rs, ont, sq, ccs. [ccs]
                         ccs: -i 0.9 -N 0.1
                   sq/rs/ont: -i 0.75 -N 0.2
   -i FLOAT      minimal sequence identity for variant match. [0.9]
   -r FLOAT      minimal ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.5]
   -N FLOAT      maximal NM ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.1]
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -q INT        minimal mapping quality [20]
                 Reads with mapping quality smaller than threshold will be ignored
   -c FLOAT      expected sampling coverage for local consensus [40], 
                 0 for no coverage sampling
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --cns-chunk-size INT
                 maximal reference chunk size to collect reads data to perform
                 local consensus [1000]. Reads of variants with reference
                 distance < INT will be collected to perform local consensus
   --cns-side-ext-size INT
                 region extend size on both sides while generating consensus sequences
                 [1000] bp.
   --cns-side-ext-size-clip INT
                 clip region extend size on both sides while generating consensus sequences
                 [5000] bp.
   --min-cons-read-size INT
                 minimal segment size for consensus [100] bp.
   --keep-cns-reads
                 Keep temporary reads from being deleted during local consensus.
                 This may take some additional disk space
   --re-cns-failed-work
                 Reperform previously failed local consensus work.
   --include-alt
                 include alt chromosomal items in result [False]
   --include-decoy
                 include decoy chromosomal items in result [False]
   --sample STR  Sample name ["sample"]
   -v,--version  show version information
   -h,--help     show this help message and exit

Example:
   # run 'cns' command on the whole genome or the user-specified regions according to previous 'det' command.
   $ asvclr cns -t 32 -n 3 -m 20 -x ccs -o output ref.fa genome_sorted.bam
```

Note that: the `cns` step can be re-run from last stop to avoid unnecessary recomputation, and the `-x` option can be used to sampling high local coverage to a relative lower coverage to accelerate assemble process if the expected sampling coverage option `-x` is specified as a positive value.

### `Call` Step

* For each variant, the consensus sequences (contigs) are aligned onto the local reference to generate the PAF format alignment with CIGAR tag by using minimap2 with '-c -x asm5' option, then the variant will be detected with their genotyes.

```sh
$ asvclr call -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr call
Program: ASVCLR (Allele-aware Structural Variant Caller for Long Reads)
Version: 1.4.6 (using htslib 1.17)

Usage: asvclr call [options] <REF_FILE> <BAM_FILE>

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted BAM file (required)

Options: 
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be max(2+floor(0.03*depth), 3).
   -x STR        sequencing technology preset: rs, ont, sq, ccs. [ccs]
                         ccs: -i 0.9 -I 0.95 -N 0.1
                   sq/rs/ont: -i 0.75 -I 0.7 -N 0.2
   -i FLOAT      minimal sequence identity for variant match. [0.9]
   -I FLOAT      minimal sequence identity for merge [0.95]
   -d INT        minimal reference distance for merge [300]
                 Neighboring SVs with distance smaller than INT and sequence identity
                 larger than FLOAT will be merged
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -q INT        minimal mapping quality [20]
                 Reads with mapping quality smaller than threshold will be ignored
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --include-alt
                 include alt chromosomal items in result [False]
   --include-decoy
                 include decoy chromosomal items in result [False]
   --sample STR  Sample name ["sample"]
   --gt-min-consist-merge FLOAT
                 minimal sequence identity for allele merge [0.95].
                 Allelic variants will be merged if their sequence identity are larger than FLOAT. 
   --gt-homo-ratio FLOAT
                 minimal allele ratio for homozygous alleles [0.75].
                 Variant is homozygous if the ratio of allele count is larger than FLOAT.
   --gt_hete_ratio FLOAT
                 minimal allele ratio for heterozygous alleles [0.2].
                 Variant is heterozygous if the ratio of allele count is larger than FLOAT.
   -v,--version  show version information
   -h,--help     show this help message and exit

Example:
   # run 'call' command on the whole genome or the user-specified regions according to previous 'det' command.
   $ asvclr call -t 32 -n 3 -m 20 -o output ref.fa genome_sorted.bam
```


## Output Result Description

Variant detection results are reported in VCF file format.

### VCF file format

Variants can be reported in VCF file format, and specifically, translocations are recorded by their breakend information which typically can be denoted as `BND` type, and a translocation variant may have 8 breakends at most, i.e. 8 `BND` items. In ASVCLR, the VCF file format can be described as below:

```sh
##fileformat=VCFv4.2
##fileDate=20240820
##source=ASVCLR 1.4.6
##reference=hs37d5.fa
##PG="asvclr all -p hg002_minimap2_hs37d5_asvclr -o output hs37d5.fa hg002_minimap2_sorted.bam"
##contig=<ID=1,length=249250621>
##contig=<ID=2,length=243199373>
##contig=<ID=3,length=198022430>
##contig=<ID=4,length=191154276>
...
##contig=<ID=NC_007605,length=171823>
##contig=<ID=hs37d5,length=35477943>
##phasing=None
##INFO=<ID=END,Number=1,Type=Integer,Description="End position of the structural variant described in this record">
##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant">
##INFO=<ID=SVLEN,Number=1,Type=Integer,Description="Difference in length between REF and ALT alleles">
##INFO=<ID=DUPNUM,Number=1,Type=Integer,Description="Copy number of DUP">
##INFO=<ID=MATEID,Number=.,Type=String,Description="ID of mate breakends">
##INFO=<ID=MATEDIST,Number=1,Type=Integer,Description="Distance to the mate breakend for mates on the same contig">
##INFO=<ID=LDISCOV,Number=1,Type=String,Description="Variant discover level: ALN for variants are discovered from consensus alignments, RES-ALN for variants are discovered from rescued consensus alignments around variant regions, READS for variants are discovered from reads alignments directly">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Read depth per allele">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read depth at this position for this sample">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample
1	32028	.	G	GTCATTCTCAGCTGGTCCTGGACCTCAGGTCCATTCTCAGCCTGGACCTCAGCTGGTCAAACCTGG	.	PASS	SVTYPE=INS;SVLEN=64;END=32028;DP=5;AF=0.8;LDISCOV=ALN	GT:AD:DP	1/1:1,4:5
1	757842	.	GCCTGGCCAGCAGATCCACCCTGTCTATACTACCTG	G	.	PASS	SVTYPE=DEL;SVLEN=35;END=757876;DP=36;AF=0.861;LDISCOV=ALN	GT:AD:DP	1/1:5,31:36
1	931648	.	GACCCCCACGTGAGGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGCCCCCCCGCGGGAAGAGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGACCCCCCGCTGGAGGGGGCACCCCACGTCTGGGGCCACAGGATGCAGGGTGGGGAGGA	GACCCCCACGTGAGGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGCCCCCCCGCGGGAAGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGCCCCCCCGCGGGAAGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGACCCCCCGCTGGAGGGGGCACCTCACGTCTGGGGCCACAGGATGCAGGGTGGGGAGGA	.	PASS	SVTYPE=DUP;SVLEN=66;END=931838;DUPNUM=1;DP=32;AF=1;LDISCOV=ALN	GT:AD:DP	1/1:0,32:32
1	26968085	.	AGCGCTGGGATTACAGGTGTGAGCCACC... 	AGCGCTGGGATTACAGGTGTGAGCCACC..	.	PASS	SVTYPE=INV;SVLEN=3;END=26972951;DP=40;AF=1;LDISCOV=ALN	GT:AD:DP	1/1:0,40:40
......
```

------------------

## Contact

If you have problems or some suggestions, please contact: xzhu@ytu.edu.cn without hesitation. 

---- Enjoy !!! -----
