# ASVCLR

```
     @@@            @@@@@         @@     @@@          @@@@@         @@@            @@@@@@  
     @@@           @@@@@@@        @@@    @@@        @@@@@@@@        @@@            @@@@@@@ 
    @@@@@          @@@            @@@   @@@         @@@             @@@            @@   @@ 
    @@@@@          @@@            @@@   @@@        @@@              @@@            @@   @@ 
   @@@ @@          @@@@            @@@ @@@         @@@              @@@            @@  @@@ 
   @@@ @@@          @@@@@          @@@ @@@         @@@              @@@            @@@@@@  
   @@@@@@@           @@@@@         @@@ @@@         @@@              @@@            @@@@@@  
  @@@@@@@@@            @@@          @@@@@          @@@              @@@            @@  @@  
  @@@   @@@        @   @@@          @@@@@           @@@   @         @@@            @@  @@@ 
 @@@    @@@        @@@@@@@           @@@            @@@@@@@@        @@@@@@@        @@   @@@
 @@@     @@@        @@@@@            @@@              @@@@@         @@@@@@@        @@   @@@
```

Accurate Structural Variant Caller for Long Reads

-------------------
ASVCLR is an Accurate Structural Variation (SV) Caller for Long Reads, such as PacBio sequencing and Oxford Nanopore sequencing. ASVCLR can detect both simple indels (insertions and deletions) and the hard-to-call complex structural varitions (SVs) (e.g. >=20 bp), such as tandem duplications, inversions and translocations, and producing fewer false positives with high precise variant margins. 
The main description of ASVCLR can be summarized as follows:
* ASVCLR detect candidate variation regions according to the variation signatures and categorizes the candidate variation region as indel regions and clipping regions. The indel regions typically have much more indel events within the align segments while the clipping regions typically have much clipping ends (soft- and hard-clipping). ASVCLR was designed to infer the variations in the indel regions and clipping regions, respectively.
* ASVCLR consists of three steps (or commands): `det`, `cns` and `call`. In `det` step, the indel regions and and clipping genomic regions are detected as candidate variation regions; and subsequently, the `cns` (consensus) step will be used to perform the consensus operation to generate the consensus sequences by using abPOA and wtdbg2 for these candidate indel regions and clipping regions, respectively; and finally, in the `call` step, these consensus sequences derived from indel regions and clipping regions will be aligned back to reference by minimap2 to infer the variation margin and sequences.
* ASVCLR extracts variation features in indel regions and clipping regions, respectively, and adopts an allele-aware clustering for these region to avoid the interfere impact of different alleles to generate more accurate variation call results.
For more detailed experiment information, please refer to [asvclr-experiments](https://github.com/zhuxiao/asvclr-experiments/).

## Prerequisites
ASVCLR depends on the following libraries and tools:
* HTSlib (http://www.htslib.org/download/)
* abPOA (https://github.com/yangao07/abPOA/releases)
* minimap2 (https://github.com/lh3/minimap2/releases)
* g++ (v4.7 or higher which supports c++11)
* wtdbg2 2.5 (https://github.com/ruanjue/wtdbg2/releases)
The above library and tools should be installed before compiling ASVCLR. abPOA, minimap2 and wtdbg2 should be globally accessible in the computer system, these executable files `abpoa`, `minimap2` and `wtdbg2` or their soft links should be included in the `$PATH` directory.


## Compiling ASVCLR

The binary file can be generated by typing:
```sh
$ wget -c https://github.com/zhuxiao/asvclr/releases/download/1.3.0/asvclr_1.3.0.tar.xz
$ tar -xf asvclr_1.3.0.tar.xz
$ cd asvclr_1.3.0/
$ ./auto_gen.sh
```
And the binary file `asvclr` will be output into the folder `bin` in this package directory.

Besides, it is recommanded to create the link of `asvclr` in `bin` directory to the `$PATH` directory.
```sh
# create soft-link of asvclr to `~/bin` which should be already exist or make a new one
$ mkdir ~/bin
$ ln -s $PWD/bin/asvclr ~/bin
# or, create soft-link of asvclr to /usr/local/bin, which needs the sudo permission
$ sudo ln -s $PWD/bin/asvclr /usr/local/bin
```


### Compiling Prerequisites

Compling prerequisites:

* Ubuntu:
```sh
$ sudo apt install g++
```

* CentOS:
```sh
$ sudo dnf install gcc-c++
```


## Quick Start

Simply, ASVCLR can be run by typing the `all` command:
```sh
$ asvclr all -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```
Then, the following commands `det`, `cns`, `call`, `all` and `det-cns` will be performed in turn.

Moreover, ASVCLR can be used to detect variations for some user-specified regions.
```sh
# call variations in user-specified regions of hg38: chr1, chr2:10000000-20000000, chr3:50000000-100000000
$ asvclr all -o out_dir hg38.fa hg38_ngmlr_sorted.bam chr1 chr2:10000000-20000000 chr3:50000000-100000000
```
Then, only the three regions `chr1`, `chr2:10000000-20000000`, `chr3:50000000-100000000` will be processed.

The help information can be shown:
```sh
Program: ASVCLR (Accurate Structural Variation Caller for Long Reads)
Version: 1.3.0 (using htslib 1.17)

Usage: asvclr all [options] <REF_FILE> <BAM_FILE> [Region ...]

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted file (required)
   Region        Reference regions to process: CHR|CHR:START-END.
                 If unspecified, all reference regions will be 
                 processed (optional)

Options: 
   -b INT        block size [1000000]
   -s INT        Slide window size [500]
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be 0.07 times of the sequencing depth of
                 the data set (rounded)
   -r FLOAT      minimal ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.5]
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -x FLOAT      expected sampling coverage for local consensus [40], 
                 0 for no coverage sampling
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --cns-chunk-size INT
                 maximal reference chunk size to collect reads data to perform
                 local consensus [1000]. Reads of variants with reference
                 distance < INT will be collected to perform local consensus
   --cns-side-ext-size INT
                 region extend size on both sides while generating consensus sequences
                 [1000] bp.
   --cns-side-ext-size-clip INT
                 clip region extend size on both sides while generating consensus sequences
                 [20000] bp.
   --min-cons-read-size INT
                 minimal read segment size to generate consensus sequences [100] bp.
   --keep-cns-reads
                 Keep temporary reads from being deleted during local consensus.
                 This may take some additional disk space
   --re-cns-failed-work
                 Reperform previously failed local consensus work.
   --technology STR
                 Sequencing technology [pacbio]:
                   pacbio     : the PacBio CLR sequencing technology;
                   nanopore   : the Nanopore sequencing technology;
                   pacbio-hifi: the PacBio CCS sequencing technology.
   --include-decoy
                 include decoy items in result
   --sample STR  Sample name ["sample"]
   --gt-min-consist-merge FLOAT
                 minimal sequence consistency threshold for allele merge [0.9].
                 Allelic variations will be merged if their sequence consistency are larger than FLOAT. 
   --gt-homo-ratio FLOAT
                 minimal allele ratio threshold for homozygous alleles [0.75].
                 Variation is homozygous if the ratio of allele count is larger than FLOAT.
   --gt_hete_ratio FLOAT
                 minimal allele ratio threshold for heterozygous alleles [0.2].
                 Variation is heterozygous if the ratio of allele count is larger than FLOAT.
   -v,--version  show version information
   -h,--help     show this help message and exit
```
where, the htslib version is the version of HTSlib installed on the machine.

Besides, the overall help information can be shown as below:

```sh
$ asvclr
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 1.3.0 (using htslib 1.17)

Usage:  asvclr <command> [options] <REF_FILE> <BAM_FILE> [Region ...]

Description:
   REF_FILE          Reference file (required)
   BAM_FILE          Coordinate-sorted BAM file (required)
   Region            Reference regions to process: CHR|CHR:START-END.
                     If unspecified, all reference regions will be 
                     processed (optional)

Commands:
   det               detect indel signatures in aligned reads
   cns               consensus candidate regions
   call              call indels by alignments of local genome assemblies
   all               run the above commands in turn
   det-cns           run 'det' and 'cns' commands in turn
```


## Usage

Alternatively, there are three steps to run ASVCLR: `det`, `cns` and `call`.

```sh
$ asvclr det -o out_dir hg38.fa hg38_ngmlr_sorted.bam
$ asvclr cns -o out_dir hg38.fa hg38_ngmlr_sorted.bam
$ asvclr call -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

The reference and an sorted BAM file will be the input of ASVCLR, and the variants stored in the VCF file format will be generated as the output.


### `Det` Step

Structural variant regions will be detected according to variant signatures. These regions includes insertions, deletions, duplications, inversions and translocations.

```sh
$ asvclr det -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr det
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 1.3.0 (using htslib 1.17)

Usage: asvclr det [options] <REF_FILE> <BAM_FILE> [Region ...]

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted BAM file (required)
   Region        Reference regions to process: CHR|CHR:START-END.
                 If unspecified, all reference regions will be 
                 processed (optional)

Options: 
   -b INT        block size [1000000]
   -s INT        Slide window size [500]
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be 0.07 times of the sequencing depth of
                 the data set (rounded)
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --min-cons-read-size INT
                 minimal read segment size to generate consensus sequences [100] bp.
   --include-decoy
                 include decoy items in result
   --sample STR  Sample name ["sample"]
   -v,--version  show version information
   -h,--help     show this help message and exit
```

### `Cns` Step

In `Cns` step, ASVCLR extracts the corresponding local reference first. And then, ASVCLR generate the contigs by different strategies.
* For Indels in the inner-alignments, ASVCLR clusters the variation signatures by the modified NW (Needleman Wunsch) pairwise alignment algorithm, and for each cluster, ASVCLR extracts the read sequences and smoothing to remove sequence errors and short variation items. Then, ASVCLR performs the consensus operation by using abPOA to generate consensus sequences.
* For variations in clipping regions, such as DUPs, INVs and TRAs, ASVCLR performs local assembly by using wtdbg2.

```sh
$ asvclr cns -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr cns
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 1.3.0 (using htslib 1.17)

Usage: asvclr cns [options] <REF_FILE> <BAM_FILE>

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted BAM file (required)

Options: 
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -n INT        minimal number of reads supporting a SV [-1].
                 When it is not specified, -1 means the value will be
                 estimated to be 0.07 times of the sequencing depth of
                 the data set (rounded)
   -r FLOAT      minimal ratio threshold of the largest split-alignment segment
                 of a read allowing for indel detection. [0.5]
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -x FLOAT      expected sampling coverage for local consensus [40], 
                 0 for no coverage sampling
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --cns-chunk-size INT
                 maximal reference chunk size to collect reads data to perform
                 local consensus [1000]. Reads of variants with reference
                 distance < INT will be collected to perform local consensus
   --cns-side-ext-size INT
                 region extend size on both sides while generating consensus sequences
                 [1000] bp.
   --cns-side-ext-size-clip INT
                 clip region extend size on both sides while generating consensus sequences
                 [20000] bp.
   --min-cons-read-size INT
                 minimal read segment size to generate consensus sequences [100] bp.
   --keep-cns-reads
                 Keep temporary reads from being deleted during local consensus.
                 This may take some additional disk space
   --re-cns-failed-work
                 Reperform previously failed local consensus work.
   --technology STR
                 Sequencing technology [pacbio]:
                   pacbio     : the PacBio CLR sequencing technology;
                   nanopore   : the Nanopore sequencing technology;
                   pacbio-hifi: the PacBio CCS sequencing technology.
   --include-decoy
                 include decoy items in result
   --sample STR  Sample name ["sample"]
   -v,--version  show version information
   -h,--help     show this help message and exit
```

Note that: the `cns` step can be re-run from last stop to avoid unnecessary recomputation, and the `-x` option can be used to sampling high local coverage to a relative lower coverage to accelerate assemble process if the expected sampling coverage option `-x` is specified as a positive value.

### `Call` Step

* For each variation, the consensus sequences (contigs) are aligned onto the local reference to generate the PAF format alignment with CIGAR tag by using minimap2 with '-c -x asm5' option, then the variation will be detected with their genotyes.

```sh
$ asvclr call -o out_dir hg38.fa hg38_ngmlr_sorted.bam
```

And the help information are shown below:

```sh
$ asvclr call
Program: asvclr (Accurate Structural Variation Caller for Long Reads)
Version: 1.3.0 (using htslib 1.17)

Usage: asvclr call [options] <REF_FILE> <BAM_FILE>

Description:
   REF_FILE      Reference file (required)
   BAM_FILE      Coordinate-sorted BAM file (required)

Options: 
   -m INT        minimal SV size to report [20]
   -M INT        maximal SV size to report [50000]
                 Variants with size smaller than threshold will be ignored
   -e INT        minimal clipping end size [200]. Clipping events
                 with size smaller than threshold will be ignored
   -o DIR        output directory [output]
   -p STR        prefix of output result files [null]
   -t INT        number of threads [0]. 0 for the maximal number
                 of threads in machine
   --include-decoy
                 include decoy items in result
   --sample STR  Sample name ["sample"]
   --gt-min-consist-merge FLOAT
                 minimal sequence consistency threshold for allele merge [0.9].
                 Allelic variations will be merged if their sequence consistency are larger than FLOAT. 
   --gt-homo-ratio FLOAT
                 minimal allele ratio threshold for homozygous alleles [0.75].
                 Variation is homozygous if the ratio of allele count is larger than FLOAT.
   --gt_hete_ratio FLOAT
                 minimal allele ratio threshold for heterozygous alleles [0.2].
                 Variation is heterozygous if the ratio of allele count is larger than FLOAT.
   -v,--version  show version information
   -h,--help     show this help message and exit

```


## Output Result Description

Variant detection results are reported in VCF file format.

### VCF file format

Variants can be reported in VCF file format, and specifically, translocations are recorded by their breakend information which typically can be denoted as `BND` type, and a translocation variant may have 8 breakends at most, i.e. 8 `BND` items. In ASVCLR, the VCF file format can be described as below:

```sh
##fileformat=VCFv4.2
##fileDate=20240130
##source=ASVCLR 1.3.0
##reference=hs37d5.fa
##PG="asvclr all -o output hs37d5.fa hg002_ngmlr_sorted.bam"
##contig=<ID=1,length=249250621>
##contig=<ID=2,length=243199373>
##contig=<ID=3,length=198022430>
##contig=<ID=4,length=191154276>
...
##contig=<ID=NC_007605,length=171823>
##contig=<ID=hs37d5,length=35477943>
##phasing=None
##INFO=<ID=END,Number=1,Type=Integer,Description="End position of the structural variant described in this record">
##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant">
##INFO=<ID=SVLEN,Number=1,Type=Integer,Description="Difference in length between REF and ALT alleles">
##INFO=<ID=DUPNUM,Number=1,Type=Integer,Description="Copy number of DUP">
##INFO=<ID=MATEID,Number=.,Type=String,Description="ID of mate breakends">
##INFO=<ID=MATEDIST,Number=1,Type=Integer,Description="Distance to the mate breakend for mates on the same contig">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Read depth per allele">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read depth at this position for this sample">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample
1	32028	.	G	GTCATTCTCAGCTGGTCCTGGACCTCAGGTCCATTCTCAGCCTGGACCTCAGCTGGTCAAACCTGG	.	PASS	SVTYPE=INS;SVLEN=64;END=32028;DP=5;AF=0.8	GT:AD:DP	1/1:1,4:5
1	757842	.	GCCTGGCCAGCAGATCCACCCTGTCTATACTACCTG	G	.	PASS	SVTYPE=DEL;SVLEN=35;END=757876;DP=36;AF=0.861	GT:AD:DP	1/1:5,31:36
1	931648	.	GACCCCCACGTGAGGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGCCCCCCCGCGGGAAGAGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGACCCCCCGCTGGAGGGGGCACCCCACGTCTGGGGCCACAGGATGCAGGGTGGGGAGGA	GACCCCCACGTGAGGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGCCCCCCCGCGGGAAGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGCCCCCCCGCGGGAAGGGGCACCCCACATCTGGGGCCACAGGATGCAGGGTGGGGAGGGCAGAAAGGACCCCCCGCTGGAGGGGGCACCTCACGTCTGGGGCCACAGGATGCAGGGTGGGGAGGA	.	PASS	SVTYPE=DUP;SVLEN=66;END=931838;DUPNUM=1;DP=32;AF=1	GT:AD:DP	1/1:0,32:32
1	26968085	.	AGCGCTGGGATTACAGGTGTGAGCCACC... 	AGCGCTGGGATTACAGGTGTGAGCCACC..	.	PASS	SVTYPE=INV;SVLEN=3;END=26972951;DP=40;AF=1	GT:AD:DP	1/1:0,40:40
1	1000001	BND.1:1000001-1:2000002	A	A[1:2000002[	.	PASS	SVTYPE=BND;MATEID=BND.1:2000002-1:1000001;MATEDIST=1000001	GT:AD:DP	./.:49,0:49
1	2000002	BND.1:2000002-1:1000001	A	]1:1000001]A	.	PASS	SVTYPE=BND;MATEID=BND.1:1000001-1:2000002;MATEDIST=1000001	GT:AD:DP	./.:49,0:49
1	1000002	BND.1:1000002-1:2000001	T	]1:2000001]T	.	PASS	SVTYPE=BND;MATEID=BND.1:2000001-1:1000002;MATEDIST=999999	GT:AD:DP	./.:40,1:41
1	2000001	BND.1:2000001-1:1000002	A	A[1:1000002[	.	PASS	SVTYPE=BND;MATEID=BND.1:1000002-1:2000001;MATEDIST=999999	GT:AD:DP	./.:40,1:41
1	1004999	BND.1:1004999-1:2005000	C	C[1:2005000[	.	PASS	SVTYPE=BND;MATEID=BND.1:2005000-1:1004999;MATEDIST=1000001	GT:AD:DP	./.:39,1:40
1	2005000	BND.1:2005000-1:1004999	T	]1:1004999]T	.	PASS	SVTYPE=BND;MATEID=BND.1:1004999-1:2005000;MATEDIST=1000001	GT:AD:DP	./.:39,1:40
1	1005000	BND.1:1005000-1:2004999	C	]1:2004999]C	.	PASS	SVTYPE=BND;MATEID=BND.1:2004999-1:1005000;MATEDIST=999999	GT:AD:DP	./.:44,1:45
1	2004999	BND.1:2004999-1:1005000	G	G[1:1005000[	.	PASS	SVTYPE=BND;MATEID=BND.1:1005000-1:2004999;MATEDIST=999999	GT:AD:DP	./.:44,1:45
```
The last 8 variant items corresponds to a translocation which locates between chr1:1000001-2005000 and chr2:2000002-2005000

------------------

## Contact

If you have problems or some suggestions, please contact: zhuxiao_hit@yeah.net without hesitation. 

---- Enjoy !!! -----
